src/core/TChem_Driver.cpp

```diff
@@ -0,0 +1,279 @@
+   _rtol_newton = rtol_newton.as<real_type>(1e-6);
+   _dtmin = dtmin.as<real_type>(1e-8);
+   _atol_time = atol_time.as<real_type>(1e-12);
+   _rtol_time = rtol_time.as<real_type>(1e-4);
+   _max_num_newton_iterations = max_num_newton_iterations.as<ordinal_type>(100);
+   _max_num_time_iterations = max_num_time_iterations.as<ordinal_type>(1e3);
+   _num_time_iterations_per_interval = num_time_iterations_per_interval.as<ordinal_type>(1e3);
+   _jacobian_interval = jacobian_interval.as<ordinal_type>(1);
+
+}
+
+/* Create the gas kinetic model from a YAML file */
+void TChem::Driver::createGasKineticModel(const std::string &chem_file) {
+  _chem_file = chem_file;
+  _kmd = KineticModelData(_chem_file);
+}
+
+/* Create the gas kinetic constant model data */
+void TChem::Driver::createGasKineticModelConstData() {
+  printf("Creating kmcd \n");
+  using interf_host_device_type = typename Tines::UseThisDevice<TChem::host_exec_space>::type;
+  _kmcd_host = TChem::createNCAR_KineticModelConstData<interf_host_device_type>(_kmd);
+  _kmcd_device = TChem::createNCAR_KineticModelConstData<device_type>(_kmd);
+  printf("Number of Species %d \n", _kmcd_host.nSpec);
+  printf("Number of Reactions %d \n", _kmcd_host.nReac);
+  printf("End creating kmcd \n");
+
+}
+
+/* Free gas kinetic model */
+void TChem::Driver::freeAll() {
+  g_tchem->freeGasKineticModel();
+}
+
+void TChem::Driver::freeGasKineticModel() {
+  _chem_file = std::string();
+  _therm_file = std::string();
+}
+
+void TChem::Driver::createStateVector() {
+  const ordinal_type len = TChem::Impl::getStateVectorSize(_kmcd_host.nSpec);
+  const ordinal_type nBatch = 1;
+  _state = real_type_2d_view_host("state dev", nBatch, len);
+  for (ordinal_type k = 0; k < len; k++) {
+    _state(0,k) = 0.0;
+  }
+}
+
+/* Get the state vector */
+auto TChem::Driver::getStateVector() {
+  auto state_at_0 = Kokkos::subview(_state, 0, Kokkos::ALL);
+  return state_at_0;
+}
+
+void TChem_getStateVector(TChem::real_type *state){
+  auto q = g_tchem->getStateVector();
+  auto len = TChem_getLengthOfStateVector();
+  for (ordinal_type k = 0; k < len; k++) {
+     state[k] = q[k];
+  }
+}
+
+/* Set the values of the state vector */
+void TChem_setStateVector(double *array){
+  g_tchem->setStateVector(array);
+}
+
+void TChem::Driver::setStateVector(double *array) {
+  auto len = TChem_getLengthOfStateVector();
+  for (ordinal_type k = 0; k < len; k++){
+     _state(0,k) = array[k];
+  }
+}
+
+/* Return species name */
+int TChem_getSpeciesName(int * index, char* result, const std::string::size_type buffer_size){
+  std::string specName = g_tchem->getSpeciesName(index);  
+  specName.copy(result, buffer_size);
+  result[specName.length()] = '\0';
+  return specName.length();
+}
+
+std::string TChem::Driver::getSpeciesName(int *index){
+  const auto speciesNamesHost = Kokkos::create_mirror_view(_kmcd_host.speciesNames);
+  Kokkos::deep_copy(speciesNamesHost, _kmcd_host.speciesNames);
```

Comment:
We do not need this Kokkos::deep_copy. `_kmcd_host.speciesNames` is located on the host.

Reply:
Removed deep_copy in 3c61a7c